edb/server/protocol/auth_ext/base.py

```diff
@@ -0,0 +1,48 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2016-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from datetime import datetime
+
+from . import data
+
+
+class BaseProvider:
+    def __init__(
+        self, name: str, client_id: str, client_secret: str, *, http_factory
+    ):
+        self.name = name
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.http_factory = http_factory
+
+    def get_code_url(self, state: str, redirect_uri: str) -> str:
+        raise NotImplementedError
+
+    async def exchange_code(self, code: str) -> str:
+        raise NotImplementedError
+
+    async def fetch_user_info(self, token: str) -> data.UserInfo:
+        raise NotImplementedError
+
+    async def fetch_emails(self, token: str) -> list[data.Email]:
+        raise NotImplementedError
+
+    def _maybe_isoformat_to_timestamp(self, value: str | None) -> int | None:
+        return (
+            int(datetime.fromisoformat(value).timestamp()) if value else None
```

Comment:
Are you rounding the timestamp intentionally here, btw?

Reply:
Actually, there isn't a strong reason to prefer `int` here since the underlying spec just says "JSON number", but I kind of interpreted the spec to be "number of seconds" -> "integer of seconds".

See: https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

Reply:
@1st1 ⚡ [`d6fb2fc` (#5891)](https://github.com/edgedb/edgedb/pull/5891/commits/d6fb2fcdaa0a7e11223a11606319af27272f7e57)